File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
        // sliding window technique
        
        let minValue = Number.MAX_SAFE_INTEGER
        let maxValue = 0;
    
        for(let i=0;i<prices.length;i++) {
            if(prices[i] < minValue) {
                minValue = prices[i];
            }
            else {
                maxValue = Math.max(maxValue, prices[i] - minValue);
            }
        }
        return maxValue;
};

File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        
        int result = 0;
        int minValue = Integer.MAX_VALUE;
        
        for(int i=0;i<prices.length;i++) {
            if(prices[i] < minValue) {
                minValue = prices[i];
            } else {
                result = Math.max(result, (prices[i] - minValue));
            }
        }
        
        return result;
    }
}